prov/opx/include/rdma/opx/fi_opx_hmem.h

```diff
@@ -83,16 +66,44 @@ enum fi_hmem_iface fi_opx_hmem_get_iface(const void *ptr,
 		return desc->attr.iface;
 	}
 
-	enum fi_hmem_iface iface = ofi_get_hmem_iface(ptr, device, NULL);
-	if (iface == FI_HMEM_CUDA && fi_opx_hmem_is_managed(ptr, iface)) {
-		*device = 0ul;
-		return FI_HMEM_SYSTEM;
-	}
-	return iface;
-#else
+	#if HAVE_CUDA
+		unsigned mem_type;
+		unsigned is_managed;
+		unsigned device_ordinal;
+
+		/* Each pointer in 'data' needs to have the same array index
+		   as the corresponding attribute in 'cuda_attributes' */
+		void *data[] = {&mem_type, &is_managed, &device_ordinal};
+
+		enum CUpointer_attribute_enum cuda_attributes[] = {
+			CU_POINTER_ATTRIBUTE_MEMORY_TYPE,
+			CU_POINTER_ATTRIBUTE_IS_MANAGED,
+			CU_POINTER_ATTRIBUTE_DEVICE_ORDINAL
+		};
+
+		CUresult cuda_rc = cuPointerGetAttributes(ARRAY_SIZE(cuda_attributes),
+							  cuda_attributes, data,
+							  (CUdeviceptr) ptr);
+
+		if (cuda_rc == CUDA_SUCCESS) {
+
+			if (mem_type == CU_MEMORYTYPE_DEVICE && !is_managed) {
+				*device = device_ordinal;
+				return FI_HMEM_CUDA;
+			}
+		} else if (cuda_rc != CUDA_ERROR_INVALID_CONTEXT) {
+			FI_DBG_TRACE(fi_opx_global.prov, FI_LOG_EP_DATA,
+				"Bad return code %hu from cuPointerGetAttributes()",
+				cuda_rc);
+		}
```

Comment:
Can this change be applied to `cuda_is_addr_valid` (in `src/hmem_cuda.c`) instead?

Reply:
@belynam will you respond?

Reply:
> Can this change be applied to `cuda_is_addr_valid` (in `src/hmem_cuda.c`) instead?

I can port the change to `cuda_is_addr_valid` to get multiple attributes at once instead of making separate calls to get each attribute one at a time. However, `cuda_is_addr_valid` does not currently check for managed/unified memory, and if we wanted it to, it's unclear to me how best to add that check and communicate back to the user the managed status without changing the existing API (perhaps not turning on the `FI_HMEM_DEVICE_ONLY` flag?).

Even if `cuda_is_addr_valid` provided all that we need, I'm not sure we'd be able to make use of it without incurring a performance hit. With the current  `ofi_get_hmem_iface()`, we'd be stacking at least two more function calls to get to `cuda_is_addr_valid()`, and also looping through the 5 other possible device types before we get to Nvidia/Cuda on the last iteration of the for-loop.

I can take a crack at it and open a separate PR to update `cuda_is_add_valid()`, and then assuming that is approved/merged, we can update OPX code to try it. But for this PR, I think we'd like to keep the code here as is.

Reply:
@j-xiong See #9844